src/util/merkleblock.rs

```diff
@@ -534,80 +535,82 @@ mod tests {
-                        .extract_matches(&mut match_txid3, &mut indexes)
-                        .unwrap();
-                    assert_ne!(merkle_root_3, merkle_root_1);
+                // Check with random subsets with inclusion chances 1, 1/2, 1/4, ..., 1/128
+                for att in 1..15 {
+                    let mut matches = vec![false; num_tx];
+                    let mut match_txid1 = vec![];
+                    for j in 0..num_tx {
+                        // Generate `att / 2` random bits
+                        let rand_bits = match att / 2 {
+                            0 => 0,
+                            bits => rng.gen::<u64>() >> (64 - bits),
+                        };
+                        let include = rand_bits == 0;
+                        matches[j] = include;
+
+                        if include {
+                            match_txid1.push(txids[j]);
+                        };
+                    }
+
+                    // Build the partial merkle tree
+                    let pmt1 = PartialMerkleTree::from_txids(&txids, &matches);
+                    let serialized = serialize(&pmt1);
+
+                    // Verify PartialMerkleTree's size guarantees
+                    let n = min(num_tx, 1 + match_txid1.len() * height);
+                    assert!(serialized.len() <= 10 + (258 * n + 7) / 8);
+
+                    // Deserialize into a tester copy
+                    let pmt2: PartialMerkleTree =
+                        deserialize(&serialized).expect("Could not deserialize own data");
+
+                    // Extract merkle root and matched txids from copy
+                    let mut match_txid2: Vec<Txid> = vec![];
+                    let mut indexes = vec![];
+                    let merkle_root_2 = pmt2
+                        .extract_matches(&mut match_txid2, &mut indexes)
+                        .expect("Could not extract matches");
+
+                    // Check that it has the same merkle root as the original, and a valid one
+                    assert_eq!(merkle_root_1, merkle_root_2);
+                    assert_ne!(merkle_root_2, TxMerkleNode::all_zeros());
+
+                    // check that it contains the matched transactions (in the same order!)
+                    assert_eq!(match_txid1, match_txid2);
+
+                    // check that random bit flips break the authentication
+                    for _ in 0..4 {
+                        let mut pmt3: PartialMerkleTree = deserialize(&serialized).unwrap();
+                        pmt3.damage(&mut rng);
+                        let mut match_txid3 = vec![];
+                        let merkle_root_3 = pmt3
+                            .extract_matches(&mut match_txid3, &mut indexes)
+                            .unwrap();
+                        assert_ne!(merkle_root_3, merkle_root_1);
+                    }
                 }
-            }
+            }).join().unwrap();
```

Comment:
I don't think this is doing what we want it to. Calling `join()` inside the loop makes the code block on the first thread. This can be verified by running 
```rust
    fn pmt_tests() {
        let tx_counts = vec![1, 4, 7, 17, 56, 100, 127, 256, 312, 513, 1000, 4095];

        for (i, _) in tx_counts.iter().enumerate() {
            eprintln!("loop: {}", i);
            thread::spawn(move || {
                thread::sleep(std::time::Duration::from_secs(3));
                panic!("force thread to panic");
            }).join().unwrap();
        }
    }
```

Which gives
```bash
loop: 0
thread '<unnamed>' panicked at 'force thread to panic', src/util/merkleblock.rs:542:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'util::merkleblock::tests::pmt_tests' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', src/util/merkleblock.rs:543:23
test util::merkleblock::tests::pmt_tests ... FAILED
```

Instead we can do something like the following:
```rust
    fn pmt_tests() {
        let tx_counts = vec![1, 4, 7, 17, 56, 100, 127, 256, 312, 513, 1000, 4095];
        let mut handles = Vec::with_capacity(tx_counts.len());

        for (i, _) in tx_counts.iter().enumerate() {
            eprintln!("loop: {}", i);
            let handle = thread::spawn(move || {
                thread::sleep(std::time::Duration::from_secs(3));
                panic!("force thread to panic");
            });
            handles.push(handle);
        }
        for handle in handles {
            handle.join().expect("thread panicked");
        }
    }
```